ADAL/src/cache/ADTokenCacheAccessor.m

```diff
@@ -213,51 +256,72 @@ - (void)updateCacheToItem:(ADTokenCacheItem *)cacheItem
         NSString* familyId = cacheItem.familyId;
         if (familyId)
         {
+            [[ADTelemetry sharedInstance] startEvent:[requestParams telemetryRequestId] eventName:@"token_cache_write"];
+            
             ADTokenCacheItem* frtItem = [multiRefreshTokenItem copy];
             NSString* fociClientId = [ADTokenCacheAccessor familyClientId:familyId];
             frtItem.clientId = fociClientId;
-            [_dataSource addOrUpdateItem:frtItem correlationId:correlationId error:nil];
+            [_dataSource addOrUpdateItem:frtItem correlationId:[requestParams correlationId] error:nil];
             SAFE_ARC_RELEASE(frtItem);
+            
+            ADTelemetryCacheEvent* event = [[ADTelemetryCacheEvent alloc] initWithName:@"token_cache_write"];
+            [event setTokenType:@"family refresh token"];
+            [[ADTelemetry sharedInstance] stopEvent:[requestParams telemetryRequestId] event:event];
+            SAFE_ARC_RELEASE(event);
         }
         SAFE_ARC_RELEASE(multiRefreshTokenItem);
     }
     
-    AD_LOG_VERBOSE_F(@"Token cache store", correlationId, @"Storing access token for resource: %@", cacheItem.resource);
-    [_dataSource addOrUpdateItem:cacheItem correlationId:correlationId error:nil];
+    AD_LOG_VERBOSE_F(@"Token cache store", [requestParams correlationId], @"Storing access token for resource: %@", cacheItem.resource);
+    [[ADTelemetry sharedInstance] startEvent:[requestParams telemetryRequestId] eventName:@"token_cache_write"];
+    
+    [_dataSource addOrUpdateItem:cacheItem correlationId:[requestParams correlationId] error:nil];
     cacheItem.refreshToken = savedRefreshToken;//Restore for the result
+    
+    ADTelemetryCacheEvent* event = [[ADTelemetryCacheEvent alloc] initWithName:@"token_cache_write"];
+    [event setTokenType:@"access token"];
+    [[ADTelemetry sharedInstance] stopEvent:[requestParams telemetryRequestId] event:event];
+    SAFE_ARC_RELEASE(event);
 }
 
 - (void)removeItemFromCache:(ADTokenCacheItem *)cacheItem
                refreshToken:(NSString *)refreshToken
-              correlationId:(NSUUID *)correlationId
+              requestParams:(ADRequestParameters*)requestParams
                       error:(ADAuthenticationError *)error
 {
     if (!cacheItem && !refreshToken)
     {
         return;
     }
     
+    [[ADTelemetry sharedInstance] startEvent:[requestParams telemetryRequestId] eventName:@"token_cache_delete_token"];
+    
     BOOL removed = NO;
     //The refresh token didn't work. We need to tombstone this refresh item in the cache.
     ADTokenCacheKey* exactKey = [cacheItem extractKey:nil];
     if (exactKey)
     {
-        ADTokenCacheItem* existing = [_dataSource getItemWithKey:exactKey userId:cacheItem.userInformation.userId correlationId:correlationId error:nil];
+        ADTokenCacheItem* existing = [_dataSource getItemWithKey:exactKey userId:cacheItem.userInformation.userId correlationId:[requestParams correlationId] error:nil];
```

Comment:
For cases where a single value is being used a lot it's best to put it into a local variable so you don't have to keep making objc msg calls. Also, it would reduce the amount of code you'd have to change :)